File: WD Content/InfoViewController.m
//
//  InfoViewController.m
//  WD Content
//
//  Created by Sergey Seitov on 08.12.13.
//  Copyright (c) 2013 Sergey Seitov. All rights reserved.
//

#import "InfoViewController.h"

NSString* const UpdateInfoNotification = @"UpdateInfoNotification";

@interface InfoViewController ()

@property (strong, nonatomic) NSMutableDictionary *info;
@property (strong, nonatomic) Node* node;
@property (strong, nonatomic) NSArray *fields;

@end

@implementation InfoViewController

- (void)setInfoForNode:(Node*)node
{
	_node = node;
	_info = [[NSMutableDictionary alloc] init];
	_fields = @[@"title",@"genre",@"director",@"release_date",@"runtime",@"cast",@"overview"];
	for (NSString* key in _fields) {
		id obj = [node.info valueForKey:key];
		if (obj) {
			[_info setObject:obj forKey:key];
		}
	}
	if (node.info.thumbnail) {
		[_info setObject:node.info.thumbnail forKey:@"thumbnail"];
	}
	if (node.info.original_title) {
		[_info setObject:node.info.original_title forKey:@"original_title"];
	}
	self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(done)];
	self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"Clear" style:UIBarButtonItemStylePlain target:self action:@selector(clear)];
	self.navigationItem.leftBarButtonItem.tintColor = [UIColor whiteColor];
	self.navigationItem.rightBarButtonItem.tintColor = [UIColor whiteColor];
}

- (void)setInfo:(NSDictionary*)info forNode:(Node*)node
{
	_node = node;
	_info = [[NSMutableDictionary alloc] init];
	_fields = @[@"title",@"genre",@"director",@"release_date",@"runtime",@"cast",@"overview"];
	for (NSString* key in _fields) {
		id obj = [info valueForKey:key];
		if (obj) {
			if ([obj isKindOfClass:[NSNumber class]]) {
				[_info setObject:[obj stringValue] forKey:key];
			} else if ([obj isKindOfClass:[NSString class]]) {
				[_info setObject:obj forKey:key];
			}
		}
	}
	if ([info objectForKey:@"thumbnail"]) {
		[_info setObject:[info objectForKey:@"thumbnail"] forKey:@"thumbnail"];
	}
	if ([info objectForKey:@"original_title"]) {
		[_info setObject:[info objectForKey:@"original_title"] forKey:@"original_title"];
	}
	self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:self action:@selector(apply)];
	self.navigationItem.rightBarButtonItem.tintColor = [UIColor whiteColor];
}

- (void)viewDidLoad
{
    [super viewDidLoad];

	self.title = @"Info";
}

- (void)done
{
	[self dismissViewControllerAnimated:YES completion:^(){}];
}

- (void)clear
{
	[[DataModel sharedInstance] clearInfoForNode:_node];
	[[NSNotificationCenter defaultCenter] postNotificationName:UpdateInfoNotification object:_node];
	[self dismissViewControllerAnimated:YES completion:^(){}];
}

- (void)apply
{
	[[DataModel sharedInstance] addInfo:_info forNode:_node];
	[[NSNotificationCenter defaultCenter] postNotificationName:UpdateInfoNotification object:_node];
	[self dismissViewControllerAnimated:YES completion:^(){}];
}

#pragma mark - Table view data source

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 7;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return 1;
}

- (NSString *)tableView:(UITableView *)tableView titleForHeaderInSection:(NSInteger)section
{
	return [_fields objectAtIndex:section];
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
	if (indexPath.section == 0) {
		return [_info objectForKey:@"thumbnail"] ? 120 : 44;
	} else {
		CGSize maximumSize = CGSizeMake(tableView.frame.size.width, 800);
		NSString *cellString = [_info objectForKey:[_fields objectAtIndex:indexPath.section]];
		UIFont *font = [UIFont fontWithName:@"HelveticaNeue" size:14];
		CGRect rect = [cellString boundingRectWithSize:maximumSize
											   options:NSStringDrawingUsesLineFragmentOrigin
											attributes:[NSDictionary dictionaryWithObject:font forKey:NSFontAttributeName]
											   context:nil];
		return rect.size.height > 44 ? rect.size.height : 44;
	}
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
	UITableViewCell *cell = [self.tableView dequeueReusableCellWithIdentifier:@"Cell"];
	if (!cell)
	{
		cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"Cell"];
	}
	cell.textLabel.text = [_info objectForKey:[_fields objectAtIndex:indexPath.section]];
	if (indexPath.section == 0) {
		if ([_info objectForKey:@"thumbnail"]) {
			cell.imageView.contentMode = UIViewContentModeScaleToFill;
			cell.imageView.image = [UIImage imageWithData:[_info objectForKey:@"thumbnail"]];
		}
		cell.detailTextLabel.text = [_info objectForKey:@"original_title"];
	}
    return cell;
}

@end

File: WD Content TV/InfoViewController.swift
//
//  InfoViewController.swift
//  WD Content
//
//  Created by Сергей Сейтов on 21.02.17.
//  Copyright © 2017 Sergey Seitov. All rights reserved.
//

import UIKit

class InfoViewController: UIViewController, UITableViewDataSource {

    @IBOutlet weak var imageView: UIImageView!
    @IBOutlet weak var infoTable: UITableView!
    @IBOutlet weak var castView: UITextView!
    @IBOutlet weak var overviewView: UITextView!
    @IBOutlet weak var castConstraint: NSLayoutConstraint!
	
	var info:[String:Any]?
	var imageBaseURL:String?
	var node:Node?
	var metainfo:MetaInfo?
	
	private var movieInfo:[String:Any]?
	private var credits:[String:Any]?

    override func viewDidLoad() {
        super.viewDidLoad()
		self.title = "Movie Info"
		
		castConstraint.constant = 0
		if info == nil && metainfo != nil {
			navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Clear", style: .plain, target: self, action: #selector(self.clearInfo))
		}
		
		if info != nil, let uid = info!["id"] as? Int {
			navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Save", style: .plain, target: self, action: #selector(self.saveInfo))
			SVProgressHUD.show(withStatus: "Load Info...")
			TMDB.sharedInstance().get(kMovieDBMovie, parameters: ["id" : "\(uid)"], block: { responseObject, error in
				if let movieInfo = responseObject as? [String:Any] {
					self.movieInfo = movieInfo
					TMDB.sharedInstance().get(kMovieDBMovieCredits, parameters: ["id" : "\(uid)"], block: { response, error in
						if let credits = response as? [String:Any] {
							self.credits = credits
						}
						self.showInfo()
						SVProgressHUD.dismiss()
					})
				} else {
					self.showInfo()
					SVProgressHUD.dismiss()
				}
			})
		}
    }
	
	override func viewDidAppear(_ animated: Bool) {
		super.viewDidAppear(animated)
		showInfo()
	}
	
	override func pressesEnded(_ presses: Set<UIPress>, with event: UIPressesEvent?) {
		if presses.first != nil && presses.first!.type == .menu {
			if metainfo != nil {
				dismiss(animated: true, completion: nil)
			} else {
				_ = navigationController?.popViewController(animated: true)
			}
		}
	}
	
	func showInfo() {
		if metainfo != nil {
			self.title = metainfo!.title
			if metainfo!.poster != nil, let url = URL(string: metainfo!.poster!) {
				imageView.sd_setImage(with: url, placeholderImage: UIImage(named: "movie"))
			} else {
				imageView.image = UIImage(named: "movie")
			}
			castView.text = metainfo!.cast
			overviewView.text = metainfo!.overview
		} else {
			self.title = info!["title"] as? String
			if let path = posterPath(), let url = URL(string: path) {
				imageView.sd_setImage(with: url, placeholderImage: UIImage(named: "movie"))
			} else {
				imageView.image = UIImage(named: "movie")
			}
			castView.text = cast()
			overviewView.text = info!["overview"] as? String
		}
		infoTable.reloadData()
		if castView.text != nil {
			let castHeight = castView.text!.heightWithConstrainedWidth(width: castView.frame.width, font: castView.font!) + 40
			let overviewHeight = overviewView.text!.heightWithConstrainedWidth(width: overviewView.frame.width, font: overviewView.font!) + 40
			let height = self.view.frame.height - castView.frame.origin.y - overviewHeight - 80
			castConstraint.constant = castHeight > height ? height : castHeight
		}
	}
	
	func clearInfo() {
		metainfo!.node!.info = nil
		Model.shared.clearInfo(metainfo!)
		NotificationCenter.default.post(name: refreshNotification, object: nil)
		dismiss(animated: true, completion: nil)
	}
	
	func saveInfo() {
		let uid = info!["id"] as? Int
		
		metainfo = Model.shared.createInfo("\(uid)")
		metainfo?.title = info!["title"] as? String
		metainfo?.overview = info!["overview"] as? String
		metainfo?.release_date = info!["release_date"] as? String
		metainfo?.poster = posterPath()
		metainfo?.runtime = runtime()
		metainfo?.rating = rating()
		metainfo?.genre = genres()
		metainfo?.cast = cast()
		metainfo?.director = director()
		
		if node!.info != nil {
			Model.shared.clearInfo(node!.info!)
		}
		node?.info = metainfo
		metainfo?.node = node
		Model.shared.saveContext()
		NotificationCenter.default.post(name: refreshNotification, object: nil)
		dismiss(animated: true, completion: nil)
	}
	
	private func posterPath() -> String? {
		if let posterPath = info!["poster_path"] as? String {
			return "\(imageBaseURL!)\(posterPath)"
		} else {
			return nil;
		}
	}
	
	private func runtime() -> String? {
		if movieInfo != nil, let runtime = movieInfo!["runtime"] as? Int {
			return "\(runtime)"
		} else {
			return nil
		}
	}
	
	private func genres() -> String? {
		if movieInfo != nil, let genresArr = movieInfo!["genres"] as? [Any] {
			var genres:[String] = []
			for item in genresArr {
				if let genreItem = item as? [String:Any], let genre = genreItem["name"] as? String {
					genres.append(genre)
				}
			}
			return genres.joined(separator: ", ")
		} else {
			return nil
		}
	}
	
	private func rating() -> String? {
		if movieInfo != nil, let popularity = movieInfo!["vote_average"] as? Double {
			return "\(popularity)"
		} else {
			return nil
		}
	}

	private func cast() -> String? {
		if credits != nil, let castArr = credits!["cast"] as? [Any] {
			var cast:[String] = []
			for item in castArr {
				if let casting = item as? [String:Any], let name = casting["name"] as? String {
					cast.append(name)
				}
			}
			return cast.joined(separator: ", ")
		} else {
			return nil
		}
	}
	
	private func director() -> String? {
		if credits != nil, let crewArr =  credits!["crew"] as? [Any] {
			var director:[String] = []
			for item in crewArr {
				if let crew = item as? [String:Any], let job = crew["job"] as? String, job == "Director", let name = crew["name"] as? String {
					director.append(name)
				}
			}
			return director.joined(separator: ", ")
		} else {
			return nil
		}
	}

	func numberOfSections(in tableView: UITableView) -> Int {
		return 1
	}
	
	func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
		return 5
	}
	
	func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
		let cell = UITableViewCell(style: .value1, reuseIdentifier: nil)
		switch indexPath.row {
		case 0:
			cell.textLabel?.text = "Director"
			cell.detailTextLabel?.text = metainfo != nil ? metainfo!.director : director()
		case 1:
			cell.textLabel?.text = "Release Date"
			let formatter = DateFormatter()
			formatter.dateFormat = "yyyy-MM-dd"
			if let text = metainfo != nil ? metainfo!.release_date : info!["release_date"] as? String, let date = formatter.date(from: text) {
				let yearFormatter = DateFormatter()
				yearFormatter.dateStyle = .long
				yearFormatter.timeStyle = .none
				cell.detailTextLabel?.text = yearFormatter.string(from: date)
			} else {
				cell.detailTextLabel?.text = ""
			}
		case 2:
			cell.textLabel?.text = "Runtime"
			if let runtime = metainfo != nil ? metainfo!.runtime : runtime() {
				cell.detailTextLabel?.text = "\(runtime) min"
			} else {
				cell.detailTextLabel?.text = ""
			}
		case 3:
			cell.textLabel?.text = "Genres"
			cell.detailTextLabel?.text = metainfo != nil ? metainfo!.genre : genres()
		case 4:
			cell.textLabel?.text = "Rating"
			cell.detailTextLabel?.text = metainfo != nil ? metainfo!.rating : rating()
		default:
			break
		}
		return cell
	}
}

extension String {
	
	func heightWithConstrainedWidth(width: CGFloat, font: UIFont) -> CGFloat {
		let constraintRect = CGSize(width: width, height: .greatestFiniteMagnitude)
		let boundingBox = self.boundingRect(with: constraintRect, options: .usesLineFragmentOrigin, attributes: [NSFontAttributeName: font], context: nil)
		
		return boundingBox.height
	}
}

